### ISSUE TITLE
Incorrect Type Narrowing on items inside of a Union of lists

### ISSUE THREAD

**Issue by qthequartermasterman** [microsoft/pyright#10434]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**
A clear and concise description of the behavior you are seeing and the expected behavior along with steps to reproduce it.

When iterating over a variable which is a union of lists, I expect the inferred type of that items to be the union of the types inside the individual list types. I include a code example below. 

When the union of lists is instantiated empty, pyright picks the first list in the union (list[A] in the example below) to be the type of the list. One consequence of this is adding B() to the list later is a type error, according to pyright. 

When the union of lists is populated at instantiation (such as union_of_lists = [B()]), pyright narrows the type (correctly) to the match its contents.

The example below is a little contrived, but instead of instantiating the list myself, in my use case, there is a function that is constructing the list at runtime from user input. The function is constrained to only return a homogeneous list, so it really is a union of lists, and not a list of unions, but pyright does not parse it correctly.

Mypy behaves as I would expect on the example during the first loop, but on the enumerate it reveals the intersection of the two types... I think it's a completely unrelated bug in mypy, though, which pyright doesn't emulate. If it's helpful to the conversation, here's the post on mypy: https://github.com/python/mypy/issues/8586#issuecomment-2867698435

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
class A:
    ...
class B:
    ...

union_of_lists: list[A] | list[B] = []
reveal_type(union_of_lists)  # Revealed type is "list[A]", expected: list[A] | list[B]
...
for item1 in union_of_lists:
    reveal_type(union_of_lists)  # Revealed type is "list[A]", expected: list[A] | list[B]
    reveal_type(item1)  # Revealed type is "A", expected A|B

for i, item2 in enumerate(union_of_lists):
    reveal_type(union_of_lists) # Revealed type is "list[A]", expected: list[A] | list[B]
    reveal_type(item2)  # Revealed type is "A", expected A | B
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

I'm running this both as part of pylance in VSCode and via the CLI, version 1.1.399

**Comment by erictraut** [comment]
Pyright is working correctly here, so this isn't a bug.

When you assign a value to a variable, its type is narrowed on assignment based on the type of the RHS expression. In this case, the RHS expression is `[]`.

When you construct a new object, its type must be established at construction time. The expression `[]` constructs a new list. The type of the resulting object can be either `list[A]` or `list[B]`, but it cannot be both. If you want pyright to assume that you are constructing an empty `list[B]` object, you can use `list[B]()` rather than `[]`.

If your intent is to construct either a `list[A]` or `list[B]` based on some condition, you can do the following:

```python
union_of_lists = list[A]() if condition else list[B]()
```

**Comment by ringohoffman** [comment]
> The type of the resulting object can be either list[A] or list[B], but it cannot be both.

Isn't that just the definition of a union type? And both would be an intersection?

I think it is surprising that a simple type declaration is being ignored, and that the way around it is to assign it in an unusual way.

Mypy does not handle it the same as pyright ([playground](https://mypy-play.net/?mypy=latest&python=3.12)):

```python
class A:
    ...
class B:
    ...

union_of_lists: list[A] | list[B] = []
reveal_type(union_of_lists)  # main.py:7: note: Revealed type is "Union[builtins.list[__main__.A], builtins.list[__main__.B]]"

```

**Comment by erictraut** [comment]
> Isn't that just the definition of a union type? 

No, a union describes a set of possible types. When an object is constructed, it cannot be "a set of possible types". It is a specific type at the time of construction.

Mypy is inconsistent in how it handles this case. Its behavior doesn't make sense to me.

```python
union_of_lists: list[A] | list[B] = []
reveal_type(union_of_lists)  # Mypy reveals "list[A] | list[B]"
union_of_lists.append(A()) # Mypy reports an error here
```

```python
union_of_lists: list[A] | list[B] = list[A]()
reveal_type(union_of_lists)  # Mypy reveals "list[A] | list[B]"
union_of_lists.append(A()) # Mypy reports an error here
```

```python
union_of_lists: list[A] | list[B]
# Place the assignment on a separate line
union_of_lists = []
reveal_type(union_of_lists)  # Mypy reveals "list[Never]" ?!?!?
union_of_lists.append(A()) # Mypy reports an error here
```

```python
union_of_lists: list[A] | list[B]
# Place the assignment on a separate line
union_of_lists = list[A]()
reveal_type(union_of_lists)  # Mypy reveals "list[A]"
union_of_lists.append(A()) # Mypy is OK with this
```

In any case, pyright is working as intended here. It performs type narrowing on assignment in a consistent manner.
